### PR TITLE
TimeSeries: Add support for negative Y and constant transform

### PR DESCRIPTION
**Pull request by dprokop**
Closes https://github.com/grafana/grafana/discussions/38202

This PR brings highly anticipated **Negative Y transform** to Time series panel, as well as **Constant transform**.

TODO:
- [x] Old graph -> time series migration
- [x] Documentation 

[Test dashboard.json](https://gist.github.com/dprokop/639f9f830ddf3d0591bebd9a2225058a)

Configuration UI (available **only** as an override):

image

UI indicators about series being transformed:

|  Tooltip | Legend  |
|---|---|
|image|image<br/>image|

Old graph vs timeseries basic comparison:

image

### CODE REVIEW

File: public/app/plugins/panel/timeseries/config.ts [hunk 1]
```diff
@@ -181,6 +182,29 @@ export function getGraphFieldConfig(cfg: GraphFieldConfig): SetFieldConfigOption
         });
 
       commonOptionsBuilder.addStackingConfig(builder, cfg.stacking, categoryStyles);
+
+      builder.addSelect({
+        category: categoryStyles,
+        name: 'Transform',
```
Comment (dprokop) [inline public/app/plugins/panel/timeseries/config.ts:188]: Not sure this name is clear enough, but can't think of anything else. Any suggestions are highly appreciated

File: public/app/plugins/panel/state-timeline/utils.ts [hunk 2]
```diff
@@ -325,8 +325,8 @@ export function mergeThresholdValues(field: Field, theme: GrafanaTheme2): Field
   const thresholdToText = new Map<Threshold, string>();
   const textToColor = new Map<string, string>();
   for (let i = 0; i < items.length; i++) {
-    thresholdToText.set(thresholds.steps[i], items[i].label);
-    textToColor.set(items[i].label, items[i].color!);
+    thresholdToText.set(thresholds.steps[i], items[i].name);
+    textToColor.set(items[i].name, items[i].color!);
```
Comment (ryantxu) [inline public/app/plugins/panel/state-timeline/utils.ts:329]: :thinking: looks like name is used here.    So we can not just remove it

**Review by ryantxu — COMMENTED** [review]
Looks good -- I am not sure about duplicate name, label, and key on the legend.   But also not sure I see a quick way to avoid it

**Comment by ryantxu** [comment]
If this is merged, is everything on the parity list supported?

**Comment by dprokop** [comment]
> Looks good -- I am not sure about duplicate name, label, and key on the legend. But also not sure I see a quick way to avoid it

Don't like it neither - thrown quickly to get a quick feedback on the feat. Will revisit tomorrow morning :)

File: packages/grafana-data/src/field/overrides/processors.ts [hunk 3]
```diff
@@ -74,6 +74,8 @@ export const valueMappingsOverrideProcessor = (
 export interface SelectFieldConfigSettings<T> {
   allowCustomValue?: boolean;
 
+  isClearable?: boolean;
```
[the remark this answers is not in the document]
Comment (dprokop) [reply packages/grafana-data/src/field/overrides/processors.ts:77]: Guess who added it.... ;)

**Review by simPod — CHANGES_REQUESTED** [review]
image

Do we need the `(neg Y)` there? I often have like 40 series and I think that it will be a noise. Series name is IMO sufficient.

**Comment by nward** [comment]
Hi @dprokop this is great work thanks for getting in to this.

The example dashboard has negative data in it - I think (at least for me!) the use case is to use positive values, but have it show below 0. This is particularly important for my use case as we select data by "highest" - if we have to translate it to negative then "highest" is no long really what we want.

One use case for this visualisation is for a network link, showing traffic either above or below 0 depending on the direction (rx vs. tx).
It is also the case that we would want to have multiple network links on the same chart, with their rx values stacked, and tx values stacked - with rx showing above 0 and tx showing below 0.

A couple of things I'm interested in / think you should include in your testing:
- 2 Series of positive value data, one positive (normal), and one negative-y (which should render below 0)
- 4 series of positive value data, 2 positive (normal), 2 negative-Y. The positive and negative stacks should be distinct, and should result in the stacked "negative-y" value rendering further below 0 than if it was not stacked. Perhaps stacks should automatically be separated based on negative/positive Y? I'm not sure of an example of when you'd want the negative-Y to be on the same stack as positive-Y, so feels like this should just be automatic - i.e. if you have a stack "a" on a positive-y value it should automatically be distinct from any series with stack "a" with negative-y.

**Comment by nward** [comment]
> Do we need the `(neg Y)` there? I often have like 40 series and I think that it will be a noise. Series name is IMO sufficient.

I agree with this - I don't think this is valuable for what people use this feature for. Perhaps showing this in the label could be a default disabled option?

**Comment by dprokop** [comment]
@nward @simPod  - Thanks a bunch for feedback. As of now, I have removed the transform indicators from the tooltip and the legend. 

@nward  - Me and @leeoniya have been discussing separate neg/pos y stacks a while ago. I have implemented it in the last commit. The gist is that within a single stacking group it will create a pseudo-group for the series that has negative Y transform applied and will result in a separate stack on the neg y-axis. See the screenshot in the PR description. I have also updated the test dashboard with a few more cases - feel free to test it out against this branch.

**Comment by dprokop** [comment]
> Looks good -- I am not sure about duplicate name, label, and key on the legend. But also not sure I see a quick way to avoid it

@ryantxu - I have reverted this change as it was only to support to transformation indicator in legend/tooltip - according to the feedback above it seems like it might be a useless thing. We can revisit it in the future

File: packages/grafana-ui/src/components/VizLegend/VizLegend.story.tsx [hunk 4]
```diff
@@ -158,6 +158,7 @@ function generateLegendItems(
   return [...new Array(numberOfSeries)].map((item, i) => {
     return {
       label: `${alphabet[i].toUpperCase()}-series`,
+      name: `${alphabet[i].toUpperCase()}-series`,
```
Comment (dprokop) [inline packages/grafana-ui/src/components/VizLegend/VizLegend.story.tsx:161]:
```suggestion
```

File: docs/sources/visualizations/time-series/_index.md [hunk 5]
```diff
@@ -35,6 +35,13 @@ Use these options to choose how to display your time series data.
 - [Graph stacked time series]({{< relref "./graph-time-series-stacking.md" >}})
 - [Graph and color schemes]({{< relref "./graph-color-scheme.md" >}})
 
+### Transform
+
+Use this option to transform series values without affecting values shown in tooltip, context menu and legend.
```
Comment (achatterjee-grafana) [inline docs/sources/visualizations/time-series/_index.md:40]:
```suggestion
Use this option to transform the series values without affecting the values shown in the tooltip, context menu, and legend.
```

File: docs/sources/visualizations/time-series/_index.md [hunk 6]
```diff
@@ -35,6 +35,13 @@ Use these options to choose how to display your time series data.
...
+
+- **Negative Y transform -** Flip the results to negative values on the y axis.
```
Comment (achatterjee-grafana) [inline docs/sources/visualizations/time-series/_index.md:42]:
```suggestion
- **Negative Y transform -** Flip the results to negative values on the Y axis.
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review.

**Comment by dprokop** [comment]
> If this is merged, is everything on the parity list supported?

The last thing missing after this is regions support (https://github.com/grafana/grafana/issues/35486)

**Comment by nward** [comment]
> @nward @simPod - Thanks a bunch for feedback. As of now, I have removed the transform indicators from the tooltip and the legend.
> 
> @nward - Me and @leeoniya have been discussing separate neg/pos y stacks a while ago. I have implemented it in the last commit. The gist is that within a single stacking group it will create a pseudo-group for the series that has negative Y transform applied and will result in a separate stack on the neg y-axis. See the screenshot in the PR description. I have also updated the test dashboard with a few more cases - feel free to test it out against this branch.

Nice work! That's really excellent. I'll see about getting this branch deployed somewhere it can talk to my real Influxdb data and see how it goes and get back to you.

**Comment by leeoniya** [comment]
> If this is merged, is everything on the parity list supported?
> @ryantxu The last thing missing after this is regions support (https://github.com/grafana/grafana/issues/35486)

and maybe https://github.com/grafana/grafana/discussions/38299?

**Review by zoltanbedi — APPROVED** [review]
Tried it out. LGTM!

**Review by oscarkilhed — APPROVED** [review]

**Comment by IR31108** [comment]
I'm in 9.2.4 and I'm unable to do this for a time series with only one data line. It behaves the same as multiplying by -1 which is not helpful in my case. Note that the axis is in the negatives which would be misleading. Any idea why or how to fix it?

image

**Pull request merged**